### ISSUE TITLE
Feature request: Get DOI from `DESCRIPTION`

### ISSUE THREAD

**Issue by roaldarbol** [dieghernan/cffr#82]
From the [Writing R Extensions manual](https://cran.r-project.org/doc/manuals/R-exts.html#The-DESCRIPTION-file), it specifies that custom fields are allowed. That should mean we can add a DOI field to the `DESCRIPTION` file. (I might be wrong, but that is how I understand the manual - see the last paragraph in the linked section)

With that in mind, it would be great if cffr could check for DOI in the description file, as that would allow one to auto-update both `CITATION.cff` and `inst/CITATION` in the same step whilst only manually updating `DESCRIPTION`.

**Comment by dieghernan** [comment]
Yes, it is possible to add specific fields to DESCRIPTION (we already do it for [keywords](https://docs.ropensci.org/cffr/articles/crosswalk.html#keywords), coming from `X-schema.org-keywords`). The issue I foresee is that the custom field should be somehow normalized across **R** developers (i.e. `DOI: XXXX-XXXX-XXXX`). 

In the case of `X-schema.org-keywords` the package **codemeta** somehow normalised this across **R** users, i.e. [more than 100 packages in CRAN](https://github.com/search?q=org%3Acran+X-schema.org-keywords+path%3ADESCRIPTION&type=code) already uses it. However I can't identify something similar for DOI ([search](https://github.com/search?q=org%3Acran+DOI%3A+path%3ADESCRIPTION&type=code)).

Do you wish to propose something in on this sense? Also, how the priority should be? In my mind it should be:

1. Custom `keys` value if provided.
2. `inst/CITATION`
3. `DESCRIPTION`
4. If none and the package is on CRAN, the auto-generated DOI on CRAN.

**Comment by roaldarbol** [comment]
I think the order you suggest makes complete sense!

I'm not sure I completely get the first question, is it about what the name should be? Or about how the DOI itself should be formatted? For the field name... I can't remember, are the field names in `DESCRIPTION` capitalisation-aware? (is `doi` different from `DOI`?). It seems that the [`CITATION` *bibentry* uses `doi`](https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#CITATION-files-1).
For the DOIs themselves, I see a couple of options:
- Let everything fly (...)
- Check whether DOIs are valid with regex expressions (e.g. see [this CrossRef post](https://web.archive.org/web/20180321212859/https://www.crossref.org/blog/dois-and-matching-regular-expressions/) and [retractcheck::check_doi](https://rdrr.io/cran/retractcheck/man/check_doi.html)). 
-  Try to ping the doi and see whether there's a match (e.g. see how [roadoi::fetch_oadoi](https://docs.ropensci.org/roadoi/) handles this)
- Maybe others I haven't thought of?

Maybe I completely misunderstood, in which case sorry :D

**Comment by dieghernan** [comment]
Let me explain, using a custom field as an example (`X-schema.org-keywords`). A custom field in DESCRIPTION can have any specification (caps / no caps is not a problem) for [example](https://github.com/dieghernan/tidyterra/blob/main/DESCRIPTION):

```
... 
Config/Needs/coverage: covr
Config/Needs/website: geodata, dieghernan/gitdevr, ragg, styler, metR,
    ggspatial, cpp11, remotes, gganimate, gifski
Config/testthat/edition: 3
Config/testthat/parallel: true
Encoding: UTF-8
LazyData: true
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.3.2
X-schema.org-keywords: r, terra, ggplot-extension, r-spatial, rspatial,
    cran, cran-r, r-package, rstats, rstats-package  

```
The think I see is that in order to implement this in **cffr** it should be an agreement of the whole R-devel community (as it is right now with `X-schema.org-keywords`, + 100 packages)  on how a new field including DOI should be specified. Some options:


```
...
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.3.2
X-schema.org-keywords: r, terra, ggplot-extension, r-spatial, rspatial,
    cran, cran-r, r-package, rstats, rstats-package  

DOI: 10.21105/joss.05751
doi: 10.21105/joss.05751
PackageDOI: 10.21105/joss.05751

```

If the field is standardized I could easily grab it. However my understanding is that the proper place on a **R** package for DOIs should be the `CITATION` file, as [the only purpose of supporting that file on CRAN](https://cran.r-project.org/doc/manuals/R-exts.html#CITATION-files-1) is exactly citing the package.

In case we agree on a standardization of the naming of the field I can implement it but I would expect a package with DOI to include it in a `CITATION` file, that is what **cffr** does right now.

**Comment by roaldarbol** [comment]
Ah, I see. I think it goes a bit back to my "what if we only had to manage metadata in *one* place?" argument (btw, super happy with the PR you made for it!). I don't see a reason that `inst/CITATION` shouldn't be able to fetch the DOI from `DESCRIPTION` - it already gets everything else from there.

Regarding naming, I would expect `DOI` to come out on top, simply because that's how it's presented [on CRAN packages](https://cran.r-project.org/web/packages/AalenJohansen/index.html), so the CRAN team must have already agreed on some consensus. I know it's not fetching it from `DESCRIPTION`, but it *is* what they use internally in their database:

``` r
tools::CRAN_package_db() |> 
  as.data.frame() |> 
  names()
#>  [1] "Package"                 "Version"                
#>  [3] "Priority"                "Depends"                
#>  [5] "Imports"                 "LinkingTo"              
#>  [7] "Suggests"                "Enhances"               
#>  [9] "License"                 "License_is_FOSS"        
#> [11] "License_restricts_use"   "OS_type"                
#> [13] "Archs"                   "MD5sum"                 
#> [15] "NeedsCompilation"        "Additional_repositories"
#> [17] "Author"                  "Authors@R"              
#> [19] "Biarch"                  "BugReports"             
#> [21] "BuildKeepEmpty"          "BuildManual"            
#> [23] "BuildResaveData"         "BuildVignettes"         
#> [25] "Built"                   "ByteCompile"            
#> [27] "Classification/ACM"      "Classification/ACM-2012"
#> [29] "Classification/JEL"      "Classification/MSC"     
#> [31] "Classification/MSC-2010" "Collate"                
#> [33] "Collate.unix"            "Collate.windows"        
#> [35] "Contact"                 "Copyright"              
#> [37] "Date"                    "Date/Publication"       
#> [39] "Description"             "Encoding"               
#> [41] "KeepSource"              "Language"               
#> [43] "LazyData"                "LazyDataCompression"    
#> [45] "LazyLoad"                "MailingList"            
#> [47] "Maintainer"              "Note"                   
#> [49] "Packaged"                "RdMacros"               
#> [51] "StagedInstall"           "SysDataCompression"     
#> [53] "SystemRequirements"      "Title"                  
#> [55] "Type"                    "URL"                    
#> [57] "UseLTO"                  "VignetteBuilder"        
#> [59] "ZipData"                 "Path"                   
#> [61] "X-CRAN-Comment"          "Published"              
#> [63] "Reverse depends"         "Reverse imports"        
#> [65] "Reverse linking to"      "Reverse suggests"       
#> [67] "Reverse enhances"        "Deadline"               
#> [69] "DOI"
```

<sup>Created on 2025-01-21 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

I guess the only question is whether CRAN now reserves this field, and what *they* do if a package is added which already contains a `DOI` field (https://neuromatch.social/@roaldarboel/113865982656997072).

**Comment by roaldarbol** [comment]
It seems that [no CRAN package has actually added either `doi` or `DOI` to the `DESCRIPTION` file](https://github.com/search?q=org%3Acran+%2F%5EDOI%2F+path%3ADESCRIPTION&type=code), so at least we wouldn't have to leave anyone hanging by making a choice. :-)

**Comment by zeileis** [comment]
Currently, CRAN does not put an explicit `DOI` field (or `Repository/CRAN/DOI` or something like that) into the `DESCRIPTION` because the DOI is easy to infer for all CRAN packages from the package's name. The same is true for the official URL. Both of these are added in the HTML for CRAN's landing pages automatically. (I'm also tagging @kurthornik in case he wants to add comments.)

Similarly, the auto-citation produced in `citation(..., auto = TRUE)` (the default if there is no `CITATION` file) has the following code (in R-devel since June):

    if(identical(meta$Repository, "CRAN")) {
        z$url <-
            sprintf("https://CRAN.R-project.org/package=%s", package)
        z$doi <- sprintf("10.32614/CRAN.package.%s", package)
    }

Thus, the canonical URL and official DOI are inserted automatically for all packages installed from CRAN. Similarly, URLs and (if available) DOIs are inserted automatically for packages installed from Bioconductor, R-Forge, GitHub, and GitLab. Revision numbers or SHAs are also included for the latter repositories where the version number alone might not be enough.

The general idea behind this is that you don't need to provide an explicit `CITATION` file in your package if all relevant information can be extracted from the `DESCRIPTION` anyway.

To obtain this auto-citation for some package named `pkg`, say, you can always use:

    citation("pkg", auto = TRUE)

Or if you have already read the meta-information from the `DESCRIPTION` you can do:

    meta <- packageDescription("pkg", lib.loc = ...)
    citation(auto = meta)

Note, however, that the `Repository` field in the `DESCRIPTION` is typically added upon publication on CRAN and is not present in the submitted source package.

**Comment by dieghernan** [comment]
See https://github.com/ropensci/cffr/issues/87#issuecomment-2608136730

**Comment by dieghernan** [comment]
> The general idea behind this is that you don't need to provide an explicit CITATION file in your package if all relevant information can be extracted from the DESCRIPTION anyway.

See https://github.com/ropensci/cffr/issues/87#issuecomment-2608136730
